feat: 로그인 후 헤더 구현

ssseeo0 opened the pull request:
## 🏷️ 이슈 번호 #50 

- close #50 

## 🧱 작업 사항
- 로그인 후 사용자 이름 + 아이콘 
- 위 클릭 시 팝오버 
- 자유게시판 추가했습니다 !!!(캡쳐본에는 없어요..) 

[탐이 없는 경우]
- 드롭다운 영역에 팀 생성하기 버튼

[팀이 있는 경우]
- 드롭다운 생성 
- 맨 처음 보이는 팀은 첫번째 인덱스에 해당하는 팀 
- 드롭다운을 열어서 클릭하는 경우 /[id]로 이동 
- 팀 추가하기 버튼 클릭 시 /addteam으로 이동


## 📸 결과물
로그인 전
![image](https://github.com/user-attachments/assets/17a7848b-4e14-42c2-8108-2be173a5e545)

로그인 후 - 팀 있는 경우 
![image](https://github.com/user-attachments/assets/d88fc726-34f7-4e59-ab2f-19d78933982a)

팀 드롭다운 클릭 시 해당 groupId 페이지로 이동(현재 페이지에 해당하는 경우 배경 색 다르게 지정) 
![image](https://github.com/user-attachments/assets/855eb8f6-41d1-4ae6-8643-a44ea2404627)

로그인 후 - 팀 없는 경우
![image](https://github.com/user-attachments/assets/013914ba-0b0b-4a9d-b3f3-6e545c9d98c0)

모바일 - 드롭다운, 팀 생성하기 버튼 없이 사이드바 여는 버튼만 있도록 지정 
![image](https://github.com/user-attachments/assets/833c1133-5a69-4f0a-a88a-f759b848a61d)


## 💬 공유 포인트 및 논의 사항
- 지금 팝오버가 children으로 받는 게 아니라 문자열로만 받고 있는 거 같은데 추후 팝오버 메뉴 클릭 시 이벤트를 어떻게 처리하면 좋을까요 ..? 
- 우선 제 쪽에는 페이지 이동이랑 로그아웃 기능이 있습니다 
- 다른 곳에서는 수정하기, 삭제하기 기능도 있어서 한 번 리팩토링이 필요하지 않을까 싶어요 !! 
- 모달은 승현님 PR 머지되면 추가하겠습니다 !! 
- 로그인, 회원가입 기능을 server action으로 변경했습니다. 작동은 똑같이 잘 되는데 여기서 server action이 적절한 게 맞겠쥬 ..?

ssseeo0 committed:
feat: 모바일 기준 구현

ssseeo0 committed:
feat: 사용자 정보 데이터 패칭

ssseeo0 committed:
feat: 사용자 닉네임 표시 및 컴포넌트 분리

ssseeo0 committed:
refactor: 로그인 상태 컴포넌트 분리

ssseeo0 committed:
refactor: 로그인 및 회원가입 server action으로 변경 및 로그인 후 네비게이션바 업데이트 되도록 수정

ssseeo0 committed:
feat: 드롭다운 팀 추가하기 버튼 추가

ssseeo0 committed:
feat: 사용자 아이콘 및 이름 클릭 시 팝오버 뜨도록 구현

ssseeo0 committed:
feat: 사용자 프로필 및 이름 클릭 시 팝오버

ssseeo0 committed:
refactor: 팀 생성하기 Link 태그로 변경

ssseeo0 committed:
refactor: 팀 생성하기 버튼 모바일에서 hidden 처리

codefug commented:
Deploy preview for _kkom-kkom_ ready!

✅ Preview
https://kkom-kkom-889gk1tgs-codefugs-projects.vercel.app

Built with commit fc778f034501f0834b9b9e61738a89873b2630d3.
This pull request is being automatically deployed with [vercel-action](https://github.com/marketplace/actions/vercel-action)

ssseeo0 committed:
feat: 자유게시판 추가

codefug commented on app/(auth)/action.ts:
저도 서버 액션에 대해서 자세히는 모르지만 이 코드는 아마 use server를 안 써도 돌아갈겁니다.

문서 보면 서버 액션이 form 제출할 때 action에 넣거나 이벤트 핸들러에 담을 때 필요한거 같은데 이게 참.. 어지러워요

근데 일단 위의 두 경우에 해당하는게 아니라 클라이언트 컴포넌트에서 form을 제출하기 위해서 쓰이는 거니깐 일단 비동기 함수처럼 쓰이는게 좋을 듯 해요

use server 지우고 밑에 두 함수 모두 /lib/apis/auth 에 담는게 좋을 것 같습니다. 클라이언트 컴포넌트로 서버 액션을 가지고 올 수 있어서 작동은 잘되는 거 같지만 멘토님이 저번에 하셨던 말씀들이 조금 불안하네요 (서버 액션은 클라이언트 > 서버 > 클라이언트로 동작할 수도 있다는 말씀)

ssseeo0 replied:
하 저 진짜 서버 액션 너무 헷갈려요 ㅋ 결과적으로 동작은 이전이나 지금이나 잘 되는데
서버 액션으로 했을 때의 장점을 잘 모르겠어요 .. 우선 전으로 다시 반영해 두겠습니다 !! 👍🏻

codefug commented on app/(auth)/login/_components/login-form.tsx:
```diff
@@ -45,14 +45,9 @@ export default function LoginForm() {
-      }
+      // NOTE - 로그인 후 랜딩으로 리다이렉트를 위해 push 헤더 업데이트를 위해 refresh
+      router.push("/");
+      router.refresh();
```
찾아보니깐 refresh를 하지 않으면 라우트가 바뀌어도 다시 서버 렌더링을 안 시키는군요 😊 배워갑니다

codefug commented on app/(free-board)/layout.tsx:
```diff
@@ -4,7 +4,7 @@ export default function Layout({
 }>) {
   return (
-    <div className="mx-auto w-full max-w-[343px] md:max-w-[696px] lg:max-w-[1200px]">
+    <div className="mx-auto w-full max-w-[343px] md:max-w-[696px] xl:max-w-[1200px]">
```
습관을 들여야겠어요... 또 생각났네 이제 습관이 되서 lg부터 눌러요..

codefug commented on app/action.ts:
```diff
@@ -0,0 +1,33 @@
+import { getCookie } from "cookies-next";
+import { cookies } from "next/headers";
+
+const getAccessToken = () => getCookie("accessToken", { cookies });
```
밖으로 안 빼시고 const accessToken = getCookie("accessToken", { cookies }) 하셔도 될 것 같아여

ssseeo0 replied:
다른 api 추가되면 함수 호출만 하게 하려고 일부로 빼둔 거긴 해요 !! 
내부에서 하는 게 나을까요 ??

codefug replied:
아 그런거면 이게 맞죠 아니면 cookie-next로 커스텀 함수를 하나 만들까여 서버면 서버쿠키 가져오고 클라이언트면 클라이언트 쿠키 가져오게

codefug commented on app/action.ts:
```diff
@@ -0,0 +1,33 @@
+        method: "GET",
+        headers: {
+          Authorization: `Bearer ${accessToken}`,
+          "Content-Type": "application/json",
```
GET일 때는 body가 없어서 application/json 안해도 된다고 봤던 거 같아여

ssseeo0 replied:
오오옹 감사합니다 !!

codefug commented on components/button/variants/index.ts:
좋습니다 버튼 스타일 추가하는게 엄청 편한거 같네요 보니깐 cva 왜 쓰는지 이제야 이해됐어요

ssseeo0 replied:
그니까요 저도 이거 추가하면서 너무 간단해서 제가 잘못하고 있는 줄 알았어요 완전 굿

codefug commented on components/dropdown/dropdown.tsx:
이 친구 훅을 쓰는데 use client가 없네여 추가해주세요 😊

codefug commented on components/dropdown/dropdown.tsx:
```diff
@@ -67,6 +69,7 @@ Dropdown.Item = Item;
 
 interface DropdownButtonProps {
   children: ReactNode;
+  styles?: string;
```
크게 상관은 없지만 className은 어떨까여

ssseeo0 replied:
좋습니두 !

codefug commented on components/header/header.tsx:
```diff
@@ -1,27 +1,16 @@
+  const isLoggedIn = hasCookie("accessToken", { cookies });
 
   return (
     <header className="fixed left-0 right-0 top-0 h-[60px] w-full bg-background-secondary px-4">
```
제 피알에 올라와 있긴 한데 z-index 관리에 대해서 다같이 얘기 해봐야 될 것 같아여 지금은 z-index가 없어서 자식이 헤더 위로 올라가는 경우가 생기더라구여

ssseeo0 replied:
아 z-indx 추가하는 걸 까묵었ㅇ어요 !!

codefug commented on components/header/side-menu.tsx:
```diff
@@ -0,0 +1,12 @@
+import Menu from "@/public/icons/gnb-menu.svg";
+
+export default function SideMenu() {
+  // TODO - 사이드 메뉴 열기
```
useCustomOverlay 쓰시면 될듯 합니당

codefug commented on lib/apis/type/index.ts:
정답은 없지만 저는 이 폴더의 타입들은 명세서에 있는 구조로만 갖고 있는게 안 헷갈릴 것 같아요 백엔드 서버에서 주는 데이터에 있는 Group, Membership이 모든 api에서 같은 구조라는 보장이 없어서... 호출하는 곳에서 분해하는 방식은 어떨까요

ssseeo0 replied:
저도 좀 고민하긴 했는데 호출하는 곳에서 별도로 타입을 지정하게 되면 같은 타입이 중복으로 선언되는 경우가 많지 않을까 싶어요 !! 전체 response에 대한 타입 지정은 지금처럼 냅두고,  

그 내부에 있는 type들은  group , member 등 스웨거 카테고리처럼 타입 폴더나 파일을 분리해서 각자 구현할 때 선언된 타입이 있으면 그걸 사용하고 아니면 타입 새로 정의해서 하는 건 어떠세요 ?!

codefug replied:
백엔드에서 내려주는 group, membership같은 데이터가 같은 이름인데 api마다 다른 데이터 구조를 갖고 있으면  문제가 생기더라구요.

전체를 확인해보진 않아서 모르지만 음.. 팀미팅 때 이야기 해봅시다

codefug reviewed (approved):
고생 많으셨어요! 작업 속도가 엄청 빠르시네요 😊 작업하시면서 많이 고민하신게 느껴졌습니다. 

너무 깐깐하게 리뷰를 적은 것 같아서 기분 나빠하실 수도 있겠다 싶은데요 머라도 도움이 되고 싶은 마음에 그런거니 너그럽게 봐주세요 ㅋㅋ.. 그만큼 틀을 잘 잡아주셔서 그런겁니다. 

코멘트 드린 부분은 한번씩 확인부탁드려요! 응원합니다. 👍

ssseeo0 commented:
> 고생 많으셨어요! 작업 속도가 엄청 빠르시네요 😊 작업하시면서 많이 고민하신게 느껴졌습니다.
> 
> 너무 깐깐하게 리뷰를 적은 것 같아서 기분 나빠하실 수도 있겠다 싶은데요 머라도 도움이 되고 싶은 마음에 그런거니 너그럽게 봐주세요 ㅋㅋ.. 그만큼 틀을 잘 잡아주셔서 그런겁니다.
> 
> 코멘트 드린 부분은 한번씩 확인부탁드려요! 응원합니다. 👍

어우 전 항상 승현님의 자세한 코드 리뷰 너무 좋습니다 멘토님한테 코드 리뷰 받는 너낌 ~~
저도 더 고민해 보고 반영해 둘게요 ㅎㅎ 화이팅화이팅~~

ssseeo0 committed:
refactor: 코드리뷰 반영

- 로그인 및 회원가입 server action에서 비동기 함수로 수정
- GET 불필요한 application/json 삭제
- 드롭다운 use client 선언
- 드롭다운 프롭 styles에서 className으로 변경
- 헤더 z-index 적용

hyunow commented on components/header/group-dropdown.tsx:
```diff
@@ -0,0 +1,77 @@
+                <div className="flex items-center gap-3">
+                  <div className="relative size-8 rounded-md">
+                    {/* TODO - 이미지 없는 경우 기본 이미지 */}
+                    <Image
+                      src={membership.group.image || hamster}
+                      alt={`${membership.group.name} 이미지`}
+                      fill
+                    />
```
ㅋㅋㅋㅋㅋㅋㅋㅋㅋ 기본이미지 햄스터인가요? rounded 속성 추가해주면 좋을 것 같아요

hyunow commented on components/header/group-dropdown.tsx:
```diff
@@ -0,0 +1,77 @@
+                    />
+                  </div>
+                  <p className="text-base font-medium text-white">
+                    {membership.group.name} 팀
```
이거 "팀"이 무조건 붙는거였군요 칠칠이팀 이름 수정해야겠어요 칠칠이로

hyunow reviewed (approved):
수고하셨습니다~~ 🌈 👍
